File: backend/app/services/exporters.py
"""Document export + import helpers.

Tiny standalone functions (no third-party deps) so they work in CI
without extra installs. We trade fidelity for simplicity — the HTML
exporter is a pragmatic markdown subset, good enough for round-trip
within DClaw Doc, not a CommonMark-compliant renderer.
"""
from __future__ import annotations

import html as html_lib
import json
import re

from app.models.document import Document


def export_markdown(doc: Document) -> str:
    title = doc.title or "Untitled"
    body = (doc.content_md or "").rstrip()
    if body:
        return f"# {title}\n\n{body}\n"
    return f"# {title}\n"


_HEADING_RE = re.compile(r"^(#{1,6})\s+(.*)$")
_BULLET_RE = re.compile(r"^[-*]\s+(.*)$")


def export_html(doc: Document) -> str:
    """Render a markdown subset (headings, bullets, paragraphs) to HTML."""
    title = doc.title or "Untitled"
    body_lines: list[str] = []
    in_list = False

    def close_list() -> None:
        nonlocal in_list
        if in_list:
            body_lines.append("</ul>")
            in_list = False

    for raw_line in (doc.content_md or "").splitlines():
        line = raw_line.rstrip()
        if not line:
            close_list()
            continue
        m = _HEADING_RE.match(line)
        if m:
            close_list()
            level = len(m.group(1))
            body_lines.append(f"<h{level}>{html_lib.escape(m.group(2))}</h{level}>")
            continue
        m = _BULLET_RE.match(line)
        if m:
            if not in_list:
                body_lines.append("<ul>")
                in_list = True
            body_lines.append(f"  <li>{html_lib.escape(m.group(1))}</li>")
            continue
        close_list()
        body_lines.append(f"<p>{html_lib.escape(line)}</p>")
    close_list()

    body = "\n".join(body_lines)
    return (
        "<!doctype html>\n<html><head>"
        f"<meta charset=\"utf-8\"><title>{html_lib.escape(title)}</title>"
        "</head><body>"
        f"<h1>{html_lib.escape(title)}</h1>\n{body}"
        "</body></html>"
    )


def export_json(doc: Document) -> str:
    return json.dumps(
        {
            "id": str(doc.id),
            "workspace_id": str(doc.workspace_id),
            "title": doc.title,
            "content_md": doc.content_md,
            "content_json": doc.content_json,
            "status": doc.status,
            "created_at": doc.created_at.isoformat() if doc.created_at else None,
            "updated_at": doc.updated_at.isoformat() if doc.updated_at else None,
        },
        indent=2,
    )


def parse_markdown_import(text: str) -> tuple[str, str]:
    """Split an uploaded markdown blob into ``(title, body)``.

    Uses the first H1 as the title if present; otherwise falls back to
    the first non-empty line.
    """
    title = ""
    body_lines: list[str] = []
    for idx, raw in enumerate(text.splitlines()):
        line = raw.rstrip()
        if not title:
            if line.startswith("# "):
                title = line[2:].strip()
                continue
            if line.strip():
                title = line.strip()
                # Don't skip — also include in body if it's not an H1.
        body_lines.append(line)
    body = "\n".join(body_lines).strip()
    return title or "Untitled", body

File: src/lib/exporters.ts
/**
 * Document export + import helpers, ported from the original DClaw Doc
 * exporters service. Dependency-free: the HTML exporter renders a pragmatic
 * markdown subset (headings, bullets, paragraphs), good enough for round-trip,
 * not a CommonMark-compliant renderer.
 */

export type ExportFormat = "md" | "html" | "json";

export type ExportableDoc = {
  id: string;
  workspaceId: string;
  title: string;
  contentMd: string | null;
  contentJson: unknown;
  status: string;
  createdAt: Date;
  updatedAt: Date;
};

function escapeHtml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

export function exportMarkdown(doc: ExportableDoc): string {
  const title = doc.title || "Untitled";
  const body = (doc.contentMd || "").replace(/\s+$/, "");
  return body ? `# ${title}\n\n${body}\n` : `# ${title}\n`;
}

const HEADING_RE = /^(#{1,6})\s+(.*)$/;
const BULLET_RE = /^[-*]\s+(.*)$/;

export function exportHtml(doc: ExportableDoc): string {
  const title = doc.title || "Untitled";
  const lines: string[] = [];
  let inList = false;
  const closeList = () => {
    if (inList) {
      lines.push("</ul>");
      inList = false;
    }
  };

  for (const raw of (doc.contentMd || "").split("\n")) {
    const line = raw.replace(/\s+$/, "");
    if (!line) {
      closeList();
      continue;
    }
    const h = HEADING_RE.exec(line);
    if (h) {
      closeList();
      const level = h[1].length;
      lines.push(`<h${level}>${escapeHtml(h[2])}</h${level}>`);
      continue;
    }
    const b = BULLET_RE.exec(line);
    if (b) {
      if (!inList) {
        lines.push("<ul>");
        inList = true;
      }
      lines.push(`  <li>${escapeHtml(b[1])}</li>`);
      continue;
    }
    closeList();
    lines.push(`<p>${escapeHtml(line)}</p>`);
  }
  closeList();

  const body = lines.join("\n");
  return (
    "<!doctype html>\n<html><head>" +
    `<meta charset="utf-8"><title>${escapeHtml(title)}</title>` +
    "</head><body>" +
    `<h1>${escapeHtml(title)}</h1>\n${body}` +
    "</body></html>"
  );
}

export function exportJson(doc: ExportableDoc): string {
  return JSON.stringify(
    {
      id: doc.id,
      workspace_id: doc.workspaceId,
      title: doc.title,
      content_md: doc.contentMd,
      content_json: doc.contentJson,
      status: doc.status,
      created_at: doc.createdAt?.toISOString() ?? null,
      updated_at: doc.updatedAt?.toISOString() ?? null,
    },
    null,
    2
  );
}

/**
 * Splits an uploaded markdown blob into [title, body]. Uses the first H1 as the
 * title if present; otherwise the first non-empty line.
 */
export function parseMarkdownImport(text: string): [string, string] {
  let title = "";
  const bodyLines: string[] = [];
  for (const raw of text.split("\n")) {
    const line = raw.replace(/\s+$/, "");
    if (!title) {
      if (line.startsWith("# ")) {
        title = line.slice(2).trim();
        continue;
      }
      if (line.trim()) {
        title = line.trim();
      }
    }
    bodyLines.push(line);
  }
  const body = bodyLines.join("\n").trim();
  return [title || "Untitled", body];
}

/** Builds TipTap/ProseMirror JSON from markdown text (paragraphs + headings). */
export function markdownToContentJson(md: string): unknown {
  const blocks = md.split(/\n\s*\n/).filter((b) => b.trim());
  const content = blocks.map((block) => {
    const h = HEADING_RE.exec(block.trim());
    if (h) {
      return {
        type: "heading",
        attrs: { level: Math.min(h[1].length, 6) },
        content: [{ type: "text", text: h[2] }],
      };
    }
    return { type: "paragraph", content: [{ type: "text", text: block.trim() }] };
  });
  return { type: "doc", content: content.length ? content : [{ type: "paragraph" }] };
}
